GHC `-package <pkg>` does not expose an installed package listed by `ghc-pkg list <pkg>`

### General summary/comments (optional)
Stack used as a script fails to find pandoc library although it has been downloaded and compiled.

### Steps to reproduce

Just run this script:

```haskell
#!/usr/bin/env stack
{- stack script --resolver lts-21.25 --package pandoc -}

import Prelude
import Text.Pandoc

main :: IO ()
main = putStrLn "Pandoc detected"
```

### Expected

I would expect the "Pandoc detected" printed in screen

### Actual

The script fails to compile with:

```text
/home/cebrian/borralodedentro/example.hs:5:1: error:
    Could not load module ‘Text.Pandoc’
    It is a member of the hidden package ‘pandoc-3.0.1’.
    You can run ‘:set -package pandoc’ to expose it.
    (Note: this unloads all the modules in the current scope.)
    Use -v (or `:set -v` in ghci) to see a list of the files searched for.
  |
5 | import Text.Pandoc
  | ^^^^^^^^^^^^^^^^^^

```

### Stack version

~~~text
❯ stack --version
Version 2.15.7, Git revision 75bb479f04755c1daf91ec8f445e17e2ba2ab7a3 x86_64 hpack-0.36.0
~~~

### Method of installation

* Via GHCup


### Platform

Linux Ubuntu 22.04

@tonicebrian, I think this is - at least, in part - an issue with GHC's `runghc`, and I have opened: https://gitlab.haskell.org/ghc/ghc/-/issues/25035.

My reasoning is that (on Windows):
~~~text
❯ stack --snapshot lts-21.21 exec -- runghc -hide-all-packages -package=base -package=pandoc example1.hs

example1.hs:6:1: error:
    Could not load module ‘Text.Pandoc’
    It is a member of the hidden package ‘pandoc-3.0.1’.
    You can run ‘:set -package pandoc’ to expose it.
    (Note: this unloads all the modules in the current scope.)
    Use -v (or `:set -v` in ghci) to see a list of the files searched for.
  |
6 | import Text.Pandoc
  | ^^^^^^^^^^^^^^^^^^
~~~

shoud not be a possible output of `runghc`. The very purpose of `-package=pandoc` is to expose `pandoc`.

--- a/ChangeLog.md
+++ b/ChangeLog.md
@@ -16,6 +16,11 @@ Other enhancements:
 
 Bug fixes:
 
+* `stack script --package <pkg-name>` now uses GHC's `-package-id` option to
+  expose the installed package, rather than GHC's `-package` option. For
+  packages with public sub-libraries, `-package <pkg>` can expose an
+  installed package other than one listed by `ghc-pkg list <pkg>`.
+
 ## v3.5.1 - 2025-03-29
 
 **Changes since v3.3.1:**

--- a/doc/commands/script_command.md
+++ b/doc/commands/script_command.md
@@ -95,6 +95,11 @@ statements in the source file. In that regard, Stack assumes that:
     The first time that Stack deduces package names from module names can take
     some time. Use the `--verbose` option to understand Stack's progress.
 
+!!! note
+
+    The installed packages of modules exposed by public sub-libraries will not
+    be deduced, because those installed packages are hidden.
+
 !!! info
 
     Certain packages are blacklisted because they expose one or more modules
@@ -115,6 +120,11 @@ single `--package` option can also refer to a list of package names, separated
 by a space or comma character. If the package is not in the snapshot, the most
 recent version in the package index (e.g. Hackage) will be obtained.
 
+In the case of a named public sub-library of a Cabal package, the required
+installed package is specified by the 'munged' package name. For example, for
+public sub-library `my-library` of Cabal package `my-package` the munged name of
+the installed package is `z-my-library-z-my-package`.
+
 If a required package is a GHC boot package, the behaviour can be complex. If
 the boot package has not been 'replaced', then it will be used in Stack's build
 plan. However, if the boot package has been 'replaced', the latest version of
@@ -127,7 +137,7 @@ packages but some snapshots may include directly some boot packages. In
 particular, some snapshots include directly `Win32` (which is a boot package on
 Windows) while others do not.
 
-!!! warning
+!!! info
 
     GHC has the concept of 'installed packages' (which differ from 'Cabal
     packages') in package databases. An installed package has a name. An
@@ -138,16 +148,17 @@ Windows) while others do not.
     installed package corresponding to a sub-library also has a `package-name`,
     which is the name of the Cabal package.
 
-    The `--package` option of `stack script` makes use of GHC's `-package`
-    option to expose an installed package. Unfortunately, the latter treats
-    `package-name` (if it exists) as if it were also the name of the installed
-    package. That means, for a Cabal package with one or more sub-libraries, the
-    GHC option `-package=<name>` cannot distinguish between (a) the installed
-    package `<name>` corresponding to the main library of Cabal package `<name>`
-    and (b) an installed package corresponding to a sub-library of that Cabal
-    package. The installed package that GHC picks to expose is indeterminate.
-    This can cause GHC to pick the wrong installed package and to report that it
-    cannot load a module because it is a member of a hidden package.
+    The `--package` option of `stack script` makes use of GHC's `-package-id`
+    option to expose an installed package, rather than its `-package` option.
+    The latter option treats `package-name` (if it exists) as if it were also
+    the name of the installed package. That means, for a Cabal package with one
+    or more sub-libraries, the GHC option `-package=<name>` cannot distinguish
+    between (a) the installed package `<name>` corresponding to the main library
+    of Cabal package `<name>` and (b) an installed package corresponding to a
+    sub-library of that Cabal package. The installed package that GHC picks to
+    expose is indeterminate. This can cause GHC to pick the wrong installed
+    package and to report that it cannot load a module because it is a member of
+    a hidden package.
 
 !!! warning
 
@@ -347,3 +358,34 @@ As module `Data.Text` is exposed by a number of packages that are included,
 directly or indirectly, in the specified snapshot (`incipit-base`,
 `incipit-core`, `relude` and `text`), `PackageImports` and `"text"` are required
 to specify which module is being imported.
+
+### Example 5
+
+Stackage snapshot LTS Haskell 23.18 specifies Cabal package `vector-0.13.2.0`
+which includes public sub-library `benchmarks-O2`. The sub-library exposes
+module `Bench.Vector.TestData.ParenTree` which exports `parenTree`. The
+following is a valid script:
+
+~~~haskell
+{- stack script
+   --snapshot lts-23.18
+   --package z-vector-z-benchmarks-O2
+-}
+{-# LANGUAGE LambdaCase #-}
+
+import Bench.Vector.TestData.ParenTree ( parenTree )
+import System.Environment ( getArgs )
+
+main :: IO ()
+main = getArgs >>= \case
+  [] -> putStrLn "An initial argument is required."
+  (arg:_) -> do
+    let n = read arg
+    if n >= 0 && even n
+      then do
+        putStrLn "A balanced binary tree structure"
+        putStrLn $ "with " <> show n <> " parentheses positions:"
+        print $ parenTree n
+      else
+        putStrLn "A positive even integer argument is required."
+~~~

--- a/src/Stack/Script.hs
+++ b/src/Stack/Script.hs
@@ -242,15 +242,20 @@ scriptCmd opts = do
               targets' <- mapM parsePackageNameThrowing targets
               pure $ Set.fromList targets'
 
+        GhcPkgExe pkg <- view $ compilerPathsL . to (.pkg)
+        let ghcPkgPath = toFilePath pkg
         unless (Set.null targetsSet) $ do
           -- Optimization: use the relatively cheap ghc-pkg list --simple-output
           -- to check which packages are installed already. If all needed
           -- packages are available, we can skip the (rather expensive) build
           -- call below.
-          GhcPkgExe pkg <- view $ compilerPathsL . to (.pkg)
           -- https://github.com/haskell/process/issues/251
-          bss <- snd <$> sinkProcessStderrStdout (toFilePath pkg)
-              ["list", "--simple-output"] CL.sinkNull CL.consume -- FIXME use the package info from envConfigPackages, or is that crazy?
+          bss <- snd <$> sinkProcessStderrStdout
+                   ghcPkgPath
+                   ["list", "--simple-output"]
+                   CL.sinkNull
+                   CL.consume
+          -- ^ FIXME use the package info from envConfigPackages, or is that crazy?
           let installed = Set.fromList
                         $ map toPackageName
                         $ words
@@ -264,14 +269,28 @@ scriptCmd opts = do
                     map (T.pack . packageNameString) $ Set.toList targetsSet
               withNewLocalBuildTargets targets $ build Nothing
 
-        let ghcArgs = concat
+        let packagesSet = Set.insert (mkPackageName "base") targetsSet
+            -- Yields 'raw' strings with trailing whitespace. Assumes that the
+            -- ghc-pkg application will find a package of the given name.
+            getRawPackageId :: PackageName -> RIO EnvConfig [ByteString]
+            getRawPackageId target = snd <$> sinkProcessStderrStdout
+              ghcPkgPath
+              ["field", packageNameString target, "id", "--simple-output"]
+              CL.sinkNull
+              CL.consume
+        rawPackageIds <- mapM getRawPackageId $ Set.toList packagesSet
+        let packageIds = words $ S8.unpack $ S8.concat $ concat rawPackageIds
+            -- ^ The use of words will eliminate whitespace between 'raw' items
+            ghcArgs = concat
               [ ["-i", "-i" ++ fromAbsDir (parent file)]
               , ["-hide-all-packages"]
               , maybeToList colorFlag
-              , map ("-package" ++)
-                  $ Set.toList
-                  $ Set.insert "base"
-                  $ Set.map packageNameString targetsSet
+                -- We use GHC's -package-id option rather than -package because
+                -- there is a bug in the latter. For packages with a public
+                -- sublibrary, -package <pkg> can expose an installed package
+                -- that is not listed by ghc-pkg list <pkg>. See:
+                -- https://gitlab.haskell.org/ghc/ghc/-/issues/25025
+              , map ("-package-id=" ++) packageIds
               , case shouldCompile of
                   SEInterpret -> []
                   SECompile -> []

